src/intervals/rounding.jl

```diff
@@ -2,74 +2,191 @@
+
     end
 
-    @eval begin
-        function ^{T<:AbstractFloat,S}(a::T, b::S, $mode1)
-            setrounding(T, $mode2) do
-                ^(a, b)
-            end
+
+    # power:
+
+    @eval function ^{S<:Real}(::RoundingType{:correct},
+                                        a::BigFloat, b::S, $mode1)
+                  setrounding(BigFloat, $mode2) do
+                      ^(a, b)
+                  end
+           end
+
+    # for correct rounding for Float64, must pass through BigFloat:
+    @eval function ^{S<:Real}(::RoundingType{:correct}, a::Float64, b::S, $mode1)
+        setprecision(BigFloat, 53) do
+            Float64(^(RoundingType{:correct}(), BigFloat(a), b, $mode2))
         end
     end
 
+    @eval ^{T<:AbstractFloat,S<:Real}(::RoundingType{:fast},
+                                a::T, b::S, $mode1) = $directed(a^b)
+
+    @eval ^{T<:AbstractFloat,S<:Real}(::RoundingType{:none},
+                                a::T, b::S, $mode1) = a^b
+
+
+    # functions not in CRlibm:
+    for f in (:sqrt, :inv, :tanh, :asinh, :acosh, :atanh)
+
+        @eval function $f{T<:AbstractFloat}(::RoundingType{:correct},
+                                            a::T, $mode1)
+                            setrounding(T, $mode2) do
+                                $f(a)
+                            end
+               end
+
+
+        @eval $f{T<:AbstractFloat}(::RoundingType{:fast},
+                                    a::T, $mode1) = $directed($f(a))
+
+        @eval $f{T<:AbstractFloat}(::RoundingType{:none},
+                                    a::T, $mode1) = $f(a)
+
 
-    for f in (:sqrt, :inv,
-            :tanh, :asinh, :acosh, :atanh)   # these functions not in CRlibm
-        @eval begin
-            function $f{T<:AbstractFloat}(a::T, $mode1)
-                setrounding(T, $mode2) do
-                    $f(a)
-                end
-            end
-        end
     end
 
+
+    # Functions defined in CRlibm
+
+    for f in CRlibm.functions
+        @eval $f{T<:AbstractFloat}(::RoundingType{:correct},
+                                a::T, $mode1) = CRlibm.$f(a, $mode2)
+
+        @eval $f{T<:AbstractFloat}(::RoundingType{:fast},
+                                    a::T, $mode1) = $directed($f(a))
+
+        @eval $f{T<:AbstractFloat}(::RoundingType{:none},
+                                    a::T, $mode1) = $f(a)
+
+    end
 end
+
+doc"""
+Set the rounding type used for all interval calculations on Julia v0.6 and above.
+"""
+function setrounding(::Type{Interval}, rounding_type::Symbol)
+
+    if rounding_type == current_rounding_type[]
+        return  # no need to redefine anything
+    end
+
+    if rounding_type ∉ (:correct, :fast, :none)
+        throw(ArgumentError("Rounding type must be one of `:correct`, `:fast`, `:none`"))
+    end
+
+    #rounding_type = :(RoundingType{$(Meta.quot(rounding_type))})
+    rounding_object = RoundingType{rounding_type}()
+
+
```

Comment:
The design is as follows: All functions are defined like `sin(::IntervalRounding{:correct}, x, ::RoundingMode)` in the main body of the file. Functions with different interval rounding types *coexist*, and are always all present.

In the function `setrounding(Interval, rounding_type)`, the functions like `sin(x, ::RoundingMode)` are defined, i.e. *without* the explicit interval rounding type. These functions get redefined when you call `setrounding(Interval, rounding_type)` again.

In Julia v0.6 (but *not* in 0.5), this will *automatically redefine all relevant functions*, in particular those used in `+(a::Interval, b::Interval)` etc., so that all interval functions will automatically work with the correct interval rounding type!

I will add the above explanation at the top of the file.